libyul/AsmAnalysis.cpp

```diff
@@ -101,24 +100,37 @@ AsmAnalysisInfo AsmAnalyzer::analyzeStrictAssertCorrect(Dialect const& _dialect,
 std::vector<YulString> AsmAnalyzer::operator()(Literal const& _literal)
 {
 	expectValidType(_literal.type, nativeLocationOf(_literal));
-	if (_literal.kind == LiteralKind::String && _literal.value.str().size() > 32)
+	bool erroneousLiteral = false;
+	if (_literal.kind == LiteralKind::String && !_literal.value.unlimited() && _literal.value.hint() && _literal.value.hint()->size() > 32)
+	{
+		erroneousLiteral = true;
 		m_errorReporter.typeError(
 			3069_error,
 			nativeLocationOf(_literal),
-			"String literal too long (" + std::to_string(_literal.value.str().size()) + " > 32)"
-		);
-	else if (_literal.kind == LiteralKind::Number && bigint(_literal.value.str()) > u256(-1))
+			"String literal too long (" + std::to_string(formatLiteral(_literal, false).size()) + " > 32)");
+	}
+	else if (_literal.kind == LiteralKind::Number && _literal.value.hint() && bigint(*_literal.value.hint()) > u256(-1))
+	{
+		erroneousLiteral = true;
 		m_errorReporter.typeError(6708_error, nativeLocationOf(_literal), "Number literal too large (> 256 bits)");
+	}
 	else if (_literal.kind == LiteralKind::Boolean)
-		yulAssert(_literal.value == "true"_yulstring || _literal.value == "false"_yulstring, "");
+	{
+		yulAssert(_literal.value.value() == true || _literal.value.value() == false);
+		if (_literal.value.hint())
+			yulAssert(*_literal.value.hint() == "true" || *_literal.value.hint() == "false");
+	}
 
 	if (!m_dialect.validTypeForLiteral(_literal.kind, _literal.value, _literal.type))
+	{
+		erroneousLiteral = true;
 		m_errorReporter.typeError(
 			5170_error,
 			nativeLocationOf(_literal),
-			"Invalid type \"" + _literal.type.str() + "\" for literal \"" + _literal.value.str() + "\"."
-		);
+			"Invalid type \"" + _literal.type.str() + "\" for literal \"" + formatLiteral(_literal, false) + "\".");
+	}
 
+	yulAssert(erroneousLiteral || validLiteral(_literal), "Invalid literal after validating it through AsmAnalyzer.");
```

Comment:
This assert will miss the case where we detected an error and `validLiteral()` is missing a check for it. Shouldn't it be:
```suggestion
	yulAssert(erroneousLiteral != validLiteral(_literal), "Invalid literal after validating it through AsmAnalyzer.");
```

Reply:
yes you are right, i've actually changed it to an xor; that way (i think) the intent is clearer that exactly one of the two should be true